### ISSUE TITLE
Fix S2589 FP: Assignment in catch block

### ISSUE THREAD

**Issue by groogiam** [SonarSource/sonar-dotnet#8368]
### Description

Rule S2589 false positive when variable set in catch block.

### Repro steps

```csharp
        Exception lastException = null;

        for (var i = 0; i < 5; i++)
        {
            try
            {
                await DoSomeWork();
                return;
            }
            catch (Exception ex)
            {
                lastException = ex;
                await Task.Delay(5000);
            }
        }

        if (lastException != null) //false positive here
        {
            LogError(lastException);
        }
```

### Expected behavior

S2589 should not be flagged on commented line above.

### Actual behavior

S2589 is raised.

### Related information

 C#/VB.NET Plugins version Sonarlint 7.4.0.80741
* Visual Studio version 2022 17.8.0
* MSBuild / dotnet version 8.0.100
* SonarScanner for .NET version (if used) 
* Operating System Windows 10 Pro

**Comment by zsolt-kolbay-sonarsource** [comment]
Thank you for reporting the issue. Confirmed as False Positive.

**Comment by groogiam** [comment]
Any update on when this will be fixed?

**Comment by groogiam** [comment]
@zsolt-kolbay-sonarsource Any update on when this will be fixed in SonarLint? Thanks.

**Comment by zsolt-kolbay-sonarsource** [comment]
After taking another look we concluded that is a True Positive: because of the `return` statement in the `try` block we can only reach the `if` statement when `lastException` is not `null`.

```csharp
Exception lastException = null;

for (var i = 0; i < 5; i++)
{
    try
    {
        await DoSomeWork();
        return;   // <--------- if DoSomeWork() doesn't throw then we exit the method and don't reach the if statement
    }
    catch (Exception ex)
    {
        lastException = ex;
        await Task.Delay(5000);
    }
}

if (lastException != null) // <--------- we only reach this line if DoSomeWork() failed 5 times, and at this point we can know that lastException is not null
{
    LogError(lastException);
}
```

I'm closing this issue. If you disagree, then please add a comment here, and I'll reopen it.

**Comment by groogiam** [comment]
Makes sense when I take a closer look. Looks like I was somehow thinking the return was a break statement. Thanks for your help.
